File: 두종/2026_05_28/디펜스_게임.java
import java.util.*;

class Solution {
    public int solution(int n, int k, int[] enemy) {
        // enemy의 길이가 최대 10^6 -> O(nlogn) 사용 가능
        // 일단 무적권을 쓰지 않고 막을 수 있는 최대치까지 전진
        // 전진한 인덱스까지의 숫자들 정렬 후 큰 값 부터 무적권 사용
        // 무적권을 1번 사용한 후에는 추가로 갈 수 있는 인덱스까지 다시 전진
        // 전진할 때마다 해당 숫자를 정렬된 배열에서 적절한 위치에 놓아야 하는데..
        // 우리에게 중요한 건 결국 정렬된 배열에서 가장 큰 숫자
        // Max Heap을 사용하면 되지 않을까?
        // 무적권을 사용할 수 있을 때까지 계속 사용후 최대 라운드 return
        
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b - a);
        int answer = enemy.length;
        
        for (int i = 0; i < enemy.length; i++) {
            if (n - enemy[i] >= 0) {
                pq.add(enemy[i]);
                n -= enemy[i];
            } 
            else {
                if (k > 0) {
                    pq.add(enemy[i]);
                    n -= enemy[i];
                    n += pq.poll();
                    k--;
                }
                else {
                    answer = i;
                    break;
                }
                
            }
        }
        
        return answer;
    }
}

File: 영진/2026.05.28/디펜스 게임.cpp
#include<bits/stdc++.h>

using namespace std;

int solution(int n, int k, vector<int> enemy) {
    
    //항상 가장 큰 값 - 가장 어려웠던 라운드를 내보낼 pq
    priority_queue<int> pq;
    int answer =0;
    
    for(int i =0; i<enemy.size(); i++){
        pq.push(enemy[i]);
        
        int e = enemy[i];

        if(n >= e){
            //라운드를 통과 할 수 있으면 통과
            answer++;
            n -=e;
            continue;
        }else if(k <= 0) break;

        //라운드를 통과할 수 없고 k가 남아있다면
        //가장 어려웠던 라운드에서 k를 사용 
        int m = pq.top();
        pq.pop();
        n += m; // 사용한 병사 복구
        n -= e; // 이번 라운드 클리어
        k--;  
        answer++;
    }
    
    return answer;
}